### ISSUE TITLE
`setenv()` is not thread safe

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#46002]
This issue spins off from https://github.com/cms-sw/cmssw/issues/44659 after the discovery that our code is calling `setenv()` in the parallel part of `cmsRun`, and concurrently to `std::getenv()` calls (https://github.com/cms-sw/cmssw/issues/44659#issuecomment-2349138317).

We need to first figure out a mechanism for setting environment variables during the serial part of `cmsRun` (Service constructors would satisfy the requirement, but we might want something else than moving all the `setenv()` code to new Services).

Then we need to migrate the existing code calling `setenv()`.

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
Here are some examples of the present `setenv()` calls
https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/PhysicsTools/TensorFlow/src/TensorFlow.cc#L90-L98
that is called from e.g.
https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauEmulator.cc#L720
https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauProducer.cc#L585
https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc#L211

----

https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/GeneratorInterface/RivetInterface/plugins/RivetAnalyzer.cc#L62

----

https://github.com/cms-sw/cmssw/blob/6e31ebe803aa3cd54ba56999036e36202a849dbc/HeterogeneousCore/MPIServices/src/MPIService.cc#L50

**Comment by makortel** [comment]
I tested that modifications to the `os.environ` in the `cfg.py` file are visible to the C++ side of `cmsRun`.

One option would be to implement the environment variable setting purely in the python side. I think the solution would have to
* be something that survives the pickling+unpickling, and gets executed in the worker node
* probably be good to be associated to the module configuration somehow, and applied only if the associated module gets used in the job
* probably be good to have access to the final module parameters
   * if implemented in python, the parameters injected by the configuration validation (in C++) would not be available

**Comment by makortel** [comment]
Actually the next step should be a more detailed analysis of the existing use cases of setting the environment variables in CMSSW code.

We should also find a way to get an understanding if any 3rd party code calls `setenv()` (in the parallel section of `cmsRun`). Possible ways could be e.g. a gdb script or LD_PRELOADable library to override `setenv()`.

**Comment by makortel** [comment]
It seems to me the most likely `setenv()` call to cause problems is the one in `tensorflow::setLogging()`
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/src/TensorFlow.cc#L90-L98

Unsafe calls to `tensorflow::setLogging()`:
* Module constructors or `initializeGlobalCache()`
  * [`HGCalConcentratorAutoEncoderImpl`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc#L64)
  * [`L1NNTauProducer::initializeGlobalCache()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/L1Trigger/Phase2L1ParticleFlow/plugins/L1NNTauProducer.cc#L84)
  * [`PFECALSuperClusterProducer::initializeGlobalCache()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoEcal/EgammaClusterProducers/src/PFECALSuperClusterProducer.cc#L56-L58)
    * Calls [`reco::SCProducerCache` constructor](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoEcal/EgammaCoreTools/interface/SCProducerCache.h#L29) that calls [`DeepSCGraphEvaluation` constructor](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoEcal/EgammaCoreTools/src/DeepSCGraphEvaluation.cc#L31) that calls the `setLogging()`.  * [`TSGForOIDNN`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoMuon/TrackerSeedGenerator/plugins/TSGForOIDNN.cc#L211)
  * [`L2TauNNProducer::initializeGlobalCache()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc#L237)
  * [`L2TauNNProducerAlpaka::initializeGlobalCache()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/RecoTauTag/HLTProducers/src/L2TauTagNNProducerAlpaka.cc#L236)
* `endLuminosityBlock()`
  * [`DTOccupancyTestML::dqmEndLuminosityBlock()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/DQM/DTMonitorClient/src/DTOccupancyTestML.cc#L128)
    * This module also loads the graph and creates the session in `dqmEndLuminosityBlock()`. It would be more efficient to load the graph and create the session in the constructor.
* `produce()`
  * [`L1NNCaloTauEmulator::produce()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauEmulator.cc#L720)
  * [`L1NNCaloTauProducer::produce()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauProducer.cc#L585)


Possibly unsafe call:
* [`testGraphLoadingCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testThreadPoolsCUDA.cc#L63)
   * The test program sets the Tensorflow to use a TBB thread pool, but calls the `setLogging()` before any inference calls.

Acceptable calls (in single-threaded test code in `PhysicsTools/TensorFlow`)
* [`testConstSessionCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testConstSessionCUDA.cc#L57)
* [`testGraphLoadingCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testGraphLoadingCUDA.cc#L57)
* [`testHelloWorld::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testHelloWorld.cc#L40)
* [`testHelloWorldCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testHelloWorldCUDA.cc#L60)
* [`testMetaGraphLoadingCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testMetaGraphLoadingCUDA.cc#L57)
* [`testSessionCacheCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testSessionCacheCUDA.cc#L56)
* [`testVisibleDevicesCUDA::test()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/PhysicsTools/TensorFlow/test/testVisibleDevicesCUDA.cc#L58)

Problems with this pattern
* `tensorflow::setLogging()` makes it very easy to add innocent-looking but unsafe code
* The setting is global rather than per module
* Different modules set different value (module construction order is unspecified, and thus the final outcome of these calls could differ from expectation)

Therefore, I would remove the `tensorflow::setLogging()`, and set the `TF_CPP_MIN_LOG_LEVEL=3` (or `2` if strongly motivated) in the `tensorflow` scram tool file. @smuzaffar, @cms-sw/ml-l2

**Comment by makortel** [comment]
`RivetAnalyzer::beginJob()` sets `RIVET_REF_PATH` and `RIVET_INFO_PATH` if they are not already set.
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/RivetInterface/plugins/RivetAnalyzer.cc#L53-L73

Would it be feasible to set `RIVET_REF_PATH` and `RIVET_INFO_PATH` in the `rivet` scram tool file? @smuzaffar, @cms-sw/generators-l2, @mseidel

**Comment by makortel** [comment]
`spu::Unzip` resets the `TMPDIR` if it is longer than 50 characters
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/SherpaInterface/src/SherpackUtilities.cc#L154-L161
Seems that this code was added in https://github.com/cms-sw/cmssw/pull/21682 to work around a "feature" of OpenMPI 2.0 and 2.1 (more information in https://github.com/cms-sw/cmssw/pull/21419, https://hypernews.cern.ch/HyperNews/CMS/get/edmFramework/3807/1/1/1.html, https://www.open-mpi.org/faq/?category=osx#startup-errors-with-open-mpi-2.0.x).

Our OpenMPI version is now 4.1.6, maybe this workaround is no longer needed? @cms-sw/generators-l2

For future reference, the `spu::Unzip` is called via
* [`SherpackFetcher::Fetch()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/SherpaInterface/src/SherpackFetcher.cc#L74)
* [`SherpaHadronizer` constructor](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/SherpaInterface/src/SherpaHadronizer.cc#L166)
* [`GeneratorFilter<HAD, DEC>` constructor](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/Core/interface/GeneratorFilter.h#L86)

**Comment by makortel** [comment]
`MPIService` constructor sets the `OMPI_MCA_pmix_server_uri` from a configuration parameter, if such parameter exists.
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/HeterogeneousCore/MPIServices/src/MPIService.cc#L46-L51

As Services are constructed in the serial part of `cmsRun`, this `setenv()` call is ok-ish.

**Comment by makortel** [comment]
`SiStripConfigDb::usingDatabase()` resets `TNS_ADMIN`
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/OnlineDB/SiStripConfigDb/src/SiStripConfigDb.cc#L260-L292
potentially modifying an existing value or a value from configuration. 

The `usingDatabase()` is called via
* [`SiStripConfigDb::openDbConnection()`](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/OnlineDB/SiStripConfigDb/src/SiStripConfigDb.cc#L115), but only if `UsingDb` configuraton parameter is set to `True`, and `UsingDbCache` is set to False
* [`SiStripConfigDb` constructor](https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/OnlineDB/SiStripConfigDb/src/SiStripConfigDb.cc#L53)

`SiStripConfigDb` is [defined as a Service](OnlineDB/SiStripConfigDb/plugins/module.cc), and since Services are constructed in the serial part of `cmsRun`, this `setenv()` call is ok-ish.

**Comment by makortel** [comment]
`main()` in `OnlineDB/EcalCondDB/test/LaserSeqToDB.cpp` sets `MESTORE`
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/OnlineDB/EcalCondDB/test/LaserSeqToDB.cpp#L1015
This program looks serial up to that point (unless ROOT would spawn threads), so is safe.

**Comment by makortel** [comment]
`CastorShowerLibraryMaker::update()` has a  commented out call
https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/SimG4CMS/ShowerLibraryProducer/plugins/CastorShowerLibraryMaker.cc#L686-L696
to set the `CASTOR_SL_PG_MAXE` environment variable. As being commented out, it is "safe", but maybe this commented out code could be removed? @cms-sw/simulation-l2

**Comment by makortel** [comment]
assign ml, generators, simulation

**Comment by makortel** [comment]
> Therefore, I would remove the `tensorflow::setLogging()`, and set the `TF_CPP_MIN_LOG_LEVEL=3` (or `2` if strongly motivated) in the `tensorflow` scram tool file. @smuzaffar, @cms-sw/ml-l2

I made a PR removing `tensorflow::setLogging()` https://github.com/cms-sw/cmssw/pull/46065

**Comment by makortel** [comment]
@smuzaffar Building on top of the `cmsTraceExceptions`, I crafted the following script to trace the calls to `setenv()` in the frameworks' parallel section (approximately)
```sh
#!/bin/bash

cat << EOF | gdb --args $@
set pagination off
set breakpoint pending on
# breakpoint between Service construction and first parallel section
break ScheduleItems::initMisc
run
# we don't need that breakpoint anymore
clear ScheduleItems::initMisc
# add additional breakpoint
break setenv
command
where
continue
end
continue
quit
EOF
```

Would it be feasible to run e.g. all runTheMatrix workflows for the default IB (on the default architecture) e.g. once a week?

**Comment by smuzaffar** [comment]
> Therefore, I would remove the tensorflow::setLogging(), and set the TF_CPP_MIN_LOG_LEVEL=3 (or 2 if strongly motivated) in the tensorflow scram tool file. @smuzaffar, @cms-sw/ml-l2

sounds reasonable to me @makortel . Any objections @cms-sw/ml-l2  ?

**Comment by smuzaffar** [comment]
> Would it be feasible to set RIVET_REF_PATH and RIVET_INFO_PATH in the rivet scram tool file? @smuzaffar, @cms-sw/generators-l2, @mseidel

right, scram can set there variables properly and we should do it. I will open cmsdist PR

**Comment by makortel** [comment]
> > Would it be feasible to run e.g. all runTheMatrix workflows for the default IB (on the default architecture) e.g. once a week?
> 
> Sure we can do that. So how will it working ? Should we just use `--command ' --prefix catch-setenv.sh' ` option for runTheMatrix ?

Maybe, I didn't really look into details. Where should we place the script? `cmsTraceExceptions` is outside of CMSSW, on the other hand this script would have mild dependence on the framework details to figure out an approximate point to start the tracing.

(in principle the script could also be extended to catch other unwanted functions, but maybe it's better to not overgeneralize)

**Comment by civanch** [comment]
@makortel , in previous releases of Geant4 a campaign was carried out to substitute setenv by std:setenv and for dramatic reduction of number of calls to std::setenv. So, Geant4 should not bring any problem of this kind. Concerning Castor shower library I will make PR soon.

**Comment by smuzaffar** [comment]
> Would it be feasible to set RIVET_REF_PATH and RIVET_INFO_PATH in the rivet scram tool file? @smuzaffar, @cms-sw/generators-l2, @mseidel

these we can set via scram global runtime hooks and that should fix the existing/old releases too. as code `if (!std::getenv("RIVET_REF_PATH")) { .... }` only calls `setenv` if the env variable is not set so seting it via global runtime hooks will make sure that `setenv` is not called. 

@makortel , should we do it globally or do you want to fix it in 14.2.X for now?

**Comment by makortel** [comment]
> > Would it be feasible to set RIVET_REF_PATH and RIVET_INFO_PATH in the rivet scram tool file? @smuzaffar, @cms-sw/generators-l2, @mseidel
> 
> these we can set via scram global runtime hooks and that should fix the existing/old releases too. as code `if (!std::getenv("RIVET_REF_PATH")) { .... }` only calls `setenv` if the env variable is not set so seting it via global runtime hooks will make sure that `setenv` is not called.
> 
> @makortel , should we do it globally or do you want to fix it in 14.2.X for now?

I'm fine with either way.

**Comment by makortel** [comment]
@civanch 

> in previous releases of Geant4 a campaign was carried out to substitute setenv by std:setenv and for dramatic reduction of number of calls to std::setenv. 

Just to clear up confusion, do you mean `getenv()` or `setenv()`? I'm asking because there is no `std::setenv()` (it is not part of C++ or C standards, but POSIX).

> So, Geant4 should not bring any problem of this kind. Concerning Castor shower library I will make PR soon.

Thanks!

**Comment by makortel** [comment]
> > Therefore, I would remove the tensorflow::setLogging(), and set the TF_CPP_MIN_LOG_LEVEL=3 (or 2 if strongly motivated) in the tensorflow scram tool file. @smuzaffar, @cms-sw/ml-l2
> 
> sounds reasonable to me @makortel . Any objections @cms-sw/ml-l2 ?

Assuming from https://github.com/cms-sw/cmssw/pull/46065#issuecomment-2363054393 that @cms-sw/ml-l2 is ok with the change, I opened a PR to cmsdist to set the environment variable https://github.com/cms-sw/cmsdist/pull/9418

**Comment by mseidel** [comment]
please stop tagging me, I'm the wrong guy

**Comment by makortel** [comment]
Apologies!

**Comment by makortel** [comment]
> > Would it be feasible to set RIVET_REF_PATH and RIVET_INFO_PATH in the rivet scram tool file? @smuzaffar, @cms-sw/generators-l2
> 
> these we can set via scram global runtime hooks and that should fix the existing/old releases too. as code `if (!std::getenv("RIVET_REF_PATH")) { .... }` only calls `setenv` if the env variable is not set so seting it via global runtime hooks will make sure that `setenv` is not called.
> 
> @makortel , should we do it globally or do you want to fix it in 14.2.X for now?

Tagging the right person this time @mseidel42 ...

**Comment by smuzaffar** [comment]
I have added https://github.com/cms-sw/cms-common/commit/0471f579747f77a5a723337eed1c55cdf95afa45 which should do what https://github.com/cms-sw/cmssw/blob/5587561cbbb5aa693b5982bf71e79fdb5a16ae06/GeneratorInterface/RivetInterface/plugins/RivetAnalyzer.cc#L53-L73 is doing i.e. if `RIVET_REF_PATH` or `RIVET_INFO_PATH` or not set then set these pointing to `rc/GeneratorInterface/RivetInterface/data` under `LOCALTOP` and `RELEASETOP`.

When deployed then this script will run for all cmssw releases and set `RIVET_REF_PATH` or `RIVET_INFO_PATH` when one will run `cmsenv`

**Comment by civanch** [comment]
@makortel , sorry, of course, getenv.

**Comment by civanch** [comment]
+simulation

**Comment by smuzaffar** [comment]
We also have `putenv` in few places https://github.com/search?q=repo%3Acms-sw%2Fcmssw+putenv+language%3AC%2B%2B&type=code

**Comment by makortel** [comment]
> We also have `putenv` in few places https://github.com/search?q=repo%3Acms-sw%2Fcmssw+putenv+language%3AC%2B%2B&type=code

I went through them, and all the ones needing attention seemed to be related to CondDB access. I opened a separate issue for them https://github.com/cms-sw/cmssw/issues/47038.

**Comment by makortel** [comment]
> > > Would it be feasible to run e.g. all runTheMatrix workflows for the default IB (on the default architecture) e.g. once a week?
> > 
> > 
> > Sure we can do that. So how will it working ? Should we just use `--command ' --prefix catch-setenv.sh' ` option for runTheMatrix ?
> 
> Maybe, I didn't really look into details. Where should we place the script? `cmsTraceExceptions` is outside of CMSSW, on the other hand this script would have mild dependence on the framework details to figure out an approximate point to start the tracing.

@smuzaffar I ended up in generalizing the gdb script (as I found function tracing to be useful also elsewhere), and opened now a PR https://github.com/cms-sw/cms-common/pull/18 to add a `cmsTraceFunction`. That script could be used to trace the `setenv()` in our parallel section along
```sh
cmsTraceFunction --startAfterFunction ScheduleItems::initMisc setenv cmsRun config.py
```

How could we enable this script to be run regularly for all matrix workflows? I suppose your `--command ' --prefix cmsTraceFunction --startAfterFunction ScheduleItems::initMisc setenv' ` for `runTheMatrix` would work. Would this test be as a separate IB that just runs the tests, or some other way? I think once a week frequency would be sufficient.

**Comment by makortel** [comment]
@smuzaffar Now that https://github.com/cms-sw/cms-common/pull/18 was merged, how could we use this script to be run regularly for all runTheMatrix workflows (along what I wrote above in https://github.com/cms-sw/cmssw/issues/46002#issuecomment-2881288931)?

**Comment by smuzaffar** [comment]
@makortel , we can try to add this for DEVEL_X IBs and see how it behaves. 
If needed we can also add an extra IB tests (for default/production arch only) which can run full runthematrix with few events (e.g. 2 or max 5)

**Comment by makortel** [comment]
> we can try to add this for DEVEL_X IBs and see how it behaves.

Sounds a good first step.

Although after reminding myself that we have/had `putenv()` calls as well, I'm wondering if I should extend the `cmsTraceFunction` script to be able break on multiple functions so we could catch both of them on a single run.

**Comment by smuzaffar** [comment]
https://github.com/cms-sw/cms-bot/pull/2534 should run DEVEL_X Ibs relvals under `cmsTraceFunction --startAfterFunction ScheduleItems::initMisc setenv` .

**Comment by smuzaffar** [comment]
@makortel , strangely all relvals passed for last DEVEL IB. Looks like the `cmsTraceFunction` is eating up the exit code of the cmsRun? I see that a relval tep2 has error like [a] but as the exit code was zero so runTheMatrix keep on running the step3 and above

[a] https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc12/CMSSW_15_1_DEVEL_X_2025-07-17-1100/pyRelValMatrixLogs/run/34034.0_TTbar_14TeV+Run4D120/step2_TTbar_14TeV+Run4D120.log#/
```
cmsRun: src/SimCalorimetry/HGCalSimProducers/plugins/HGCDigitizer.cc:615: void HGCDigitizer::accumulate(const edm::Handle<std::vector<PCaloHit> >&, int, const HGCalGeometry*, CLHEP::HepRandomEngine*): Assertion `waferThickness - 1 < static_cast<int>(tdcForToAOnset.size())' failed.

Thread 5 "cmsRun" received signal SIGABRT, Aborted.
[Switching to Thread 0x1554ff9ff700 (LWP 1751809)]
0x0000155553c7752f in raise () from /lib64/libc.so.6
(gdb) A debugging session is active.

	Inferior 1 [process 1750136] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
```

**Comment by makortel** [comment]
Ok, I'll see if I could make `gdb` propagate the exit code from the application.

**Comment by makortel** [comment]
https://github.com/cms-sw/cms-common/pull/19 makes `cmsTraceFunction` to propagate the exit code of the traced program. It also adds an option to trace additional functions, so  both setenv and putenv could be traced simultaneously with
```
cmsTraceFunction --startAfterFunction ScheduleItems::initMisc -f putenv setenv cmsRun config.py
```
